### PR TITLE
Add SNI and ALPN suppor for tls layer #61

### PR DESCRIPTION
**Pull request by mike820324**
This PR add several changes in order to support sni and alpn.
- Implement MicroProxyIOStream and MicroProxySSLIOStream.
- Use pyOpenSSL and service-identity for better ssl control.
- Refactor layer design.

### CODE REVIEW

**Comment by mike820324** [comment]
The coverage has dropped about 10 percent, mostly because of the following files.
- microproxy.iostream.
- microproxy.layer.application.tls

File: microproxy/context.py [hunk 1]
```diff
@@ -2,13 +2,13 @@ class Context(object):
     def __init__(self,
                  src_stream=None,
                  dest_stream=None,
+                 schema=None,
```
Comment (mike820324) [inline microproxy/context.py]: crap, I forgot to change the name into scheme.

File: microproxy/layer/application/nontls.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    def __init__(self, context):
+        super(NonTlsLayer, self).__init__()
+        self.context = copy(context)
+
+    @gen.coroutine
+    def process(self):
+        # we are not going through tls layer
+        # chage dest_stream to iostream
+        self.context.dest_stream.setblocking(False)
+        self.context.dest_stream = iostream.IOStream(self.context.dest_stream)
```
Comment (mike820324) [inline microproxy/layer/application/nontls.py:17]: Do we need to use MicroProxyIOStream or just IOStream.

File: microproxy/proxy.py [hunk 3]
```diff
@@ -14,15 +14,48 @@
...
+                logger.debug("Enter {0} Layer".format(current_layer))
+                current_context = yield current_layer.process()
+                logger.debug("Leave {0} Layer".format(current_layer))
+                layer_constructor = self.next_layer(current_layer, current_context)
+
+                if not layer_constructor:
+                    logger.debug("Layer Loop Ended")
+                    break
+            except gen.TimeoutError:
+                self.src_stream.close()
```
Comment (mike820324) [inline microproxy/proxy.py:46]: We need to break the loop when encounter exception.

**Comment by chhsiao90** [comment]
-11.3
XD

**Comment by mike820324** [comment]
ya, too many new code, I will add new test case regarding to these new code.
:dancer:

**Comment by mike820324** [comment]
Another thing about mac. When installing pyopenssl you need to specify the LDFLAGS and CFLAGS so that the crytography module can install properly with the newest openssl version.

```
LDFLAGS="/path/to/brew/lib" CFLAGS="/path/to/brew/include" pip install -r requirements
```

And another problem that I encounter in osx is the order of the pyopenssl callback function.
In linux, the `src_sni_callback` will be called before `src_alpn_callback`. However in osx, it seems like the `src_alpn_callback` will be called and as a result cause problem.

Fortunately, we call still get the servername in alpn callback by calling, 

```
src_ssl_conn.get_servername()
```

**Comment by chhsiao90** [comment]
Please rename the method process.
It didn't just do process.
It will return context.

**Comment by mike820324** [comment]
How bout  `process_and_return_context()`, which should have enough information telling the user that it will process and return a context.

**Comment by chhsiao90** [comment]
Yeah! It's good!

**Comment by mike820324** [comment]
I'll commit later. :)

**Pull request merged**